File: cpp/0051H-N皇后.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-28 16:07:25
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    vector<vector<string>> ans;

    /*
     * 递归遍历行
     * 循环遍历列
     */
    void dfs(int row, int n, vector<string>& chess) {
        if (row == n) {
            ans.push_back(chess);
            return;
        }
        for (int col = 0; col < n; col++) {
            if (check(row, col, n, chess)) {
                chess[row][col] = 'Q';
                dfs(row + 1, n, chess);
                chess[row][col] = '.';
            }
        }
    }

    /*
     *不能同行（递归过程中进行了同行检查）
     *不能同列
     *不能同斜线 （45度和135度角）
     */
    bool check(int row, int col, int n, vector<string>& chess) {
        // 检查列，剪枝
        for (int i = 0; i < row; i++) {
            if (chess[i][col] == 'Q') return false;
        }
        // ∠45°
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
            if (chess[i][j] == 'Q') return false;
        }
        // ∠135°
        for (int i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++) {
            if (chess[i][j] == 'Q') return false;
        }
        return true;
    }

    vector<vector<string>> solveNQueens(int n) {
        vector<string> chess(n, string(n, '.'));
        dfs(0, n, chess);
        return ans;
    }
};

File: redo/0051H-N皇后.go
/*
 * @Author: SukiEva
 * @Date: 2022-02-09 11:24:41
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

import (
	"strings"
)

func check(row, col, n int, chess [][]byte) bool {
	for i := 0; i < row; i++ { // 检查列
		if chess[i][col] == 'Q' {
			return false
		}
	}
	for i, j := row-1, col+1; i >= 0 && j < n; i, j = i-1, j+1 { // 45°
		if chess[i][j] == 'Q' {
			return false
		}
	}
	for i, j := row-1, col-1; i >= 0 && j >= 0; i, j = i-1, j-1 { // 135°
		if chess[i][j] == 'Q' {
			return false
		}
	}
	return true
}

func dfs(row, n int, chess [][]byte) {
	if row == n {
		t := make([]string, len(chess))
		for i, s := range chess {
			t[i] = string(s)
		}
		ans = append(ans, []string(t))
		return
	}
	for col := 0; col < n; col++ {
		if check(row, col, n, chess) {
			chess[row][col] = 'Q'
			dfs(row+1, n, chess)
			chess[row][col] = '.'
		}
	}
}

var ans [][]string

func solveNQueens(n int) [][]string {
	chess := make([][]byte, n)
	ans = [][]string{}
	for i := range chess {
		chess[i] = []byte(strings.Repeat(".", n))
	}
	dfs(0, n, chess)
	return ans
}
